src/libclipper/include/clipper/task_executor.hpp

```diff
@@ -353,23 +415,38 @@ class TaskExecutor {
   void on_response_recv(rpc::RPCResponse response) {
     std::unique_lock<std::mutex> l(inflight_messages_mutex_);
     auto keys = inflight_messages_[response.first];
+    boost::shared_lock<boost::shared_mutex> metrics_lock(model_metrics_mutex_);
 
     inflight_messages_.erase(response.first);
     std::vector<float> deserialized_outputs =
         deserialize_outputs(response.second);
     assert(deserialized_outputs.size() == keys.size());
     int batch_size = keys.size();
     predictions_counter->increment(batch_size);
-    throughput_meter->mark(batch_size);
+    throughput_meter_->mark(batch_size);
     long current_time = std::chrono::duration_cast<std::chrono::milliseconds>(
                             std::chrono::system_clock::now().time_since_epoch())
                             .count();
-    for (int batch_num = 0; batch_num < batch_size; ++batch_num) {
-      long send_time = std::get<0>(keys[batch_num]);
-      latency_hist->insert(static_cast<int64_t>(current_time - send_time));
-      cache_.put(std::get<1>(keys[batch_num]), std::get<2>(keys[batch_num]),
-                 Output{deserialized_outputs[batch_num],
-                        {std::get<1>(keys[batch_num])}});
+    if (keys.size() > 0) {
+      const VersionedModelId &cur_model = std::get<1>(keys[0]);
+      auto cur_model_metric_entry = model_metrics_.find(cur_model);
+      if (cur_model_metric_entry != model_metrics_.end()) {
```

Comment:
This seems like a good place for an `optional`. Instead of checking the predicate `cur_model_metric_entry != model_metrics_.end()` and accessing `model_metrics_entry->second` in two places, we could populate an optional that contains the model metric if it's available and check for the presence of the model metric within the optional in the inner loop. This would be more readable.